### PR TITLE
[CALCITE-4655] Fix NullPointerExeption in JdbcTable.fieldClasses

### PR DESCRIPTION
**Pull request by kramerul**
See https://issues.apache.org/jira/browse/CALCITE-4655

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/adapter/jdbc/JdbcTable.java [hunk 1]
```diff
@@ -130,7 +130,7 @@
 
   private List<Pair<ColumnMetaData.Rep, Integer>> fieldClasses(
       final JavaTypeFactory typeFactory) {
-    final RelDataType rowType = requireNonNull(protoRowType, "protoRowType").apply(typeFactory);
+    final RelDataType rowType = getRowType(typeFactory);
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/adapter/jdbc/JdbcTable.java:133]: Just for the reference, the previous code was `protoRowType.apply(typeFactory)` (see https://github.com/apache/calcite/blob/2f2fc406f4675a9ca16e3c0a352cfa4f2a1bc047/core/src/main/java/org/apache/calcite/adapter/jdbc/JdbcTable.java#L130), so nullness verifier was right that `protoRowType` should not be used like that.

I added `requireNonNull` just to avoid unrelated changes.

**Comment by kramerul** [comment]
I tried to change it to the `Supplier` Pattern, but now I get

```
[methodref.receiver.bound.invalid] Incompatible receiver type
    this.protoRowTypeSupplier = Suppliers.memoize(this::supplyProto);
                                                  ^
  found   : @UnderInitialization(org.apache.calcite.adapter.java.AbstractQueryableTable.class) @NonNull JdbcTable
  required: @Initialized @NonNull JdbcTable
  Consequence: method
    @UnderInitialization(org.apache.calcite.adapter.java.AbstractQueryableTable.class) @NonNull JdbcTable
  is not a valid method reference for
    @Initialized @NonNull RelProtoDataType supplyProto(@Initialized @NonNull JdbcTable this) in org.apache.calcite.adapter.jdbc.JdbcTable
```

I don't know how to fix this.

**Comment by vlsi** [comment]
`@SuppressWarnings("methodref.receiver.bound.invalid")` on the field declaration.

What the warning says is: "you publish a reference to uninitialized `this`", so the callee might observe unitialized object and fail with NPE or something. In other words, typically it is a bad practice to publish `this` from the middle of the constructor, however, `memoize` is a reasonable exception.

File: core/src/main/java/org/apache/calcite/adapter/jdbc/JdbcTable.java [hunk 2]
```diff
@@ -91,6 +94,7 @@
     this.jdbcSchemaName = jdbcSchemaName;
     this.jdbcTableName = requireNonNull(jdbcTableName, "jdbcTableName");
     this.jdbcTableType = requireNonNull(jdbcTableType, "jdbcTableType");
+    this.protoRowTypeSupplier = Suppliers.memoize(this::supplyProto);
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/adapter/jdbc/JdbcTable.java:97]: Frankly speaking, I would prefer to have it right at the field declaration.
